src/function/table/arrow_conversion.cpp

```diff
@@ -120,55 +120,124 @@ static void ColumnArrowToDuckDBDictionary(Vector &vector, ArrowArray &array, Arr
                                           idx_t size, const ArrowType &arrow_type, int64_t nested_offset = -1,
                                           ValidityMask *parent_mask = nullptr, uint64_t parent_offset = 0);
 
-static void ArrowToDuckDBList(Vector &vector, ArrowArray &array, ArrowArrayScanState &array_state, idx_t size,
-                              const ArrowType &arrow_type, int64_t nested_offset, ValidityMask *parent_mask,
-                              int64_t parent_offset) {
-	auto size_type = arrow_type.GetSizeType();
-	idx_t list_size = 0;
-	auto &scan_state = array_state.state;
+namespace {
 
-	SetValidityMask(vector, array, scan_state, size, parent_offset, nested_offset);
+struct ArrowListOffsetData {
+	idx_t list_size = 0;
 	idx_t start_offset = 0;
+};
+
+} // namespace
+
+template <class BUFFER_TYPE>
+static ArrowListOffsetData ConvertArrowListOffsetsTemplated(Vector &vector, ArrowArray &array, idx_t size,
+                                                            idx_t effective_offset) {
+	ArrowListOffsetData result;
+	auto &start_offset = result.start_offset;
+	auto &list_size = result.list_size;
+
 	idx_t cur_offset = 0;
+	auto offsets = ArrowBufferData<BUFFER_TYPE>(array, 1) + effective_offset;
+	start_offset = offsets[0];
+	auto list_data = FlatVector::GetData<list_entry_t>(vector);
+	for (idx_t i = 0; i < size; i++) {
+		auto &le = list_data[i];
+		le.offset = cur_offset;
+		le.length = offsets[i + 1] - offsets[i];
+		cur_offset += le.length;
+	}
+	list_size = offsets[size];
+	list_size -= start_offset;
+	return result;
+}
+
+template <class BUFFER_TYPE>
+static ArrowListOffsetData ConvertArrowListViewOffsetsTemplated(Vector &vector, ArrowArray &array, idx_t size,
+                                                                idx_t effective_offset) {
+	ArrowListOffsetData result;
+	auto &start_offset = result.start_offset;
+	auto &list_size = result.list_size;
+
+	list_size = 0;
+	auto offsets = ArrowBufferData<BUFFER_TYPE>(array, 1) + effective_offset;
+	auto sizes = ArrowBufferData<BUFFER_TYPE>(array, 2) + effective_offset;
+
+	// In ListArrays the offsets have to be sequential
+	// ListViewArrays do not have this same constraint
+	// for that reason we need to keep track of the lowest offset, so we can skip all the data that comes before it
+	// when we scan the child data
+	auto lowest_offset = offsets[0];
+	auto list_data = FlatVector::GetData<list_entry_t>(vector);
+	for (idx_t i = 0; i < size; i++) {
+		auto &le = list_data[i];
+		le.offset = offsets[i];
+		le.length = sizes[i];
+		list_size += le.length;
+		lowest_offset = MinValue(lowest_offset, offsets[i]);
+	}
+	start_offset = lowest_offset;
+	// We start scanning the child data at the 'start_offset' so we need to fix up the created list entries
+	for (idx_t i = 0; i < size; i++) {
+		auto &le = list_data[i];
+		le.offset -= start_offset;
```

Comment:
If you take the route of skipping nulls and size[i]==0 elements when calculating the `lowest_offset` and highest `offset+size` (I did, you don't have to), be very careful with these operations to ensure the invariants described on the spec (they apply to entries that are `null` as well) [1]:

```cpp
0 <= offsets[i] <= length of the child array
0 <= offsets[i] + size[i] <= length of the child array
```

[1] https://arrow.apache.org/docs/format/Columnar.html#listview-layout

Reply:
Getting this right in concatenation code was a pain. I wrote many unit tests on both Go and C++ implementations.

Reply:
That's also a good point, null entries are likely a problem because their data could be bogus right?
I wonder why that isn't an issue in `ConvertArrowListOffsetsTemplated`

Reply:
I've skipped the empty lists, and only apply the offset shifting if `lowest_offset` is non-zero
That should help with the common case 👍

Reply:
The consequences of skipping nulls and empty when calculating the `lowest_offset` is that you have to watch out when you do `offset[i] -= lowest_offset` as well to not go negative because `lowest_offset` could be greater than the values you would find on null entries.

Your approach is totally fine and avoids all this complexity. Just wanted to share highlight this nuanced aspect of the format.

> Or actually, they're not bogus?

Exactly. I started the spec saying they would be undefined/bogus, but we ended up agreeing that requiring that they should be at least valid references (hence the two invariants) would be (1) more safe, and (2) allow some compute kernels to load values at offsets without checking validity bits first. Sometimes wasted work is better than brancing inside a hot loop.